src/main/kotlin/graphics/scenery/DFTParser.kt

```diff
@@ -0,0 +1,132 @@
+    /** Number of gridpoints in 3D grid.*/
+    var gridDimensions = IntArray(3) { 0 }
+    /** Positions of the atoms in Bohr.*/
+    var atomicPositions = Array(0){ Vector3f()}
+    private var electronicDensity = Array(0, {Array(0, { FloatArray(0) } ) } )
+    /** Electronic density as scaled bytes, in scaled(1/Bohr).*/
+    lateinit var electronicDensityUInt : ByteBuffer
+    /** Origin of the unit cell, usually 0,0,0. */
+    var unitCellOrigin = FloatArray(3) { 0.0f }
+    /** Dimensions of the unit cell, in Bohr.*/
+    var unitCellDimensions = FloatArray(3) { 0.0f }
+
+    // Parse information as .cube file.
+    fun parseCube(filename: String){
+        // Read entire file content
+        val cubeFile =  File(filename).bufferedReader().readLines()
+        var counter = 0
+        var xcounter = 0
+        var ycounter = 0
+        var zcounter = 0
+        var minDensity = 1000.0f
+        var maxDensity = 0.0f
+
+        // Iterate through file. We know what is were with a cube file.
+        // 0-1: Comments.
+        // 2: Number of atoms + origin.
+        // 3,4,5: Grid spacing in x,y,z direction (in Bohr)
+        // 6 - (number_of_atoms+6): One line for each atom with position and species
+        // Everything thereafter: the volumetric data, here the density.
+
+        for (line in cubeFile){
+            when (counter){
+                0,1 ->{}
+                2 -> {
+                    numberOfAtoms = (line.trim().split("\\s+".toRegex())[0]).toInt()
+                    unitCellOrigin[0] = (line.trim().split("\\s+".toRegex())[1]).toFloat()
+                    unitCellOrigin[1] = (line.trim().split("\\s+".toRegex())[1]).toFloat()
+                    unitCellOrigin[2] = (line.trim().split("\\s+".toRegex())[1]).toFloat()
+
+                    // Now we know how many atoms we have.
+                    atomicPositions = Array(numberOfAtoms){ Vector3f()}
+                }
+                3,4,5 -> {
+                    gridDimensions[counter-3] =  (line.trim().split("\\s+".toRegex())[0]).toInt()
+                    gridSpacings[counter-3] = (line.trim().split("\\s+".toRegex())[counter-2]).toFloat()
+                }
+                else->  {
+                    if (counter == 6)
+                    {
+                        unitCellDimensions[0] = (gridSpacings[0]*gridDimensions[0])+unitCellOrigin[0]
+                        unitCellDimensions[1] = (gridSpacings[1]*gridDimensions[1])+unitCellOrigin[1]
+                        unitCellDimensions[2] = (gridSpacings[2]*gridDimensions[2])+unitCellOrigin[2]
+                        electronicDensity = Array(gridDimensions[0], { Array(gridDimensions[1],
+                            { FloatArray(gridDimensions[2]) } ) } )
+                    }
+                    // Parsing atomic positions.
+                    if (counter < 6+numberOfAtoms){
+                        atomicPositions[counter-6] = Vector3f((line.trim().split("\\s+".toRegex())[2]).toFloat(),
```

Comment:
Regex splitting is rather slow, if you're not splitting on super-special stuff, just consider using whitespace or sth.

Reply:
Also, you are repeatedly using `line.trim.split()` -- that's most likely your bottleneck, as I am not sure if the compiler has enough information available in order to optimise it out. I'd just split once per-line if possible.

Reply:
The problem is that .cube files are not necessarily consistent in their usage of whitespaces:

> In particular, the Gaussian specification is ambiguous about whitespace requirements, so parsing of CUBE files SHOULD accommodate some variation in the format, including (i) variable amounts/types of whitespace between the values on a given line, and (ii) the presence of leading and/or trailing whitespace on a given line.

https://h5cube-spec.readthedocs.io/en/latest/cubeformat.html#overview

Which is why I went with the Regex splitting. But apart from that you're absolutely right -  I was doing that way too often!